Corona-Warn-App/src/main/java/de/rki/coronawarnapp/presencetracing/risk/CheckInMatching.kt

```diff
@@ -0,0 +1,75 @@
+package de.rki.coronawarnapp.presencetracing.risk
+
+import de.rki.coronawarnapp.eventregistration.checkins.CheckIn
+import de.rki.coronawarnapp.eventregistration.checkins.download.TraceTimeIntervalWarningPackage
+import de.rki.coronawarnapp.server.protocols.internal.pt.TraceWarning
+import de.rki.coronawarnapp.util.TimeAndDateExtensions.toLocalDate
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.Deferred
+import kotlinx.coroutines.async
+import kotlinx.coroutines.withContext
+import org.joda.time.Duration
+import org.joda.time.Instant
+import kotlin.coroutines.CoroutineContext
+
+suspend fun launchMatching(
+    checkIns: List<CheckIn>,
+    warningPackages: List<TraceTimeIntervalWarningPackage>,
+    coroutineContext: CoroutineContext
+): Collection<Deferred<List<CheckInWarningOverlap>>> {
+    val launcher: CoroutineScope.(
+        List<CheckIn>,
+        TraceTimeIntervalWarningPackage
+    ) -> Deferred<List<CheckInWarningOverlap>> =
+        { list, warningPackage ->
+            async {
+                findMatches(list, warningPackage)
+            }
+        }
+
+    return warningPackages.map { warningPackage ->
+        withContext(context = coroutineContext) {
+            launcher(checkIns, warningPackage)
+        }
+    }
+}
+
+internal suspend fun findMatches(
+    checkIns: List<CheckIn>,
+    warningPackage: TraceTimeIntervalWarningPackage
+): List<CheckInWarningOverlap> {
+    return warningPackage
+        .extractTraceTimeIntervalWarnings()
+        .flatMap { warning ->
+            checkIns
+                .mapNotNull {
```

Comment:
The `mapNotNull` is a crucial part of the matching calculation, maybe a test for this?